Implement custom slice creators

EskiMojo14 opened the pull request:
see https://github.com/reduxjs/redux-toolkit/issues/3837

EskiMojo14 committed:
Custom slice creator implementation

phryneas commented on packages/toolkit/src/tsHelpers.ts:
```diff
@@ -205,3 +205,7 @@ export type Id<T> = { [K in keyof T]: T[K] } & {}
   ? Tail
   : never
+
+export type KeysMatching<T, V> = {
```
```suggestion
export type KeysForValueOfType<T, V> = {
```
`KeysMatching` sounded to me like `keyof T & keyof V`

phryneas commented on packages/toolkit/tsconfig.base.json:
```diff
@@ -32,18 +35,37 @@
-      "@reduxjs/toolkit/query": ["src/query/index.ts"], // @remap-prod-remove-line
-      "@reduxjs/toolkit/query/react": ["src/query/react/index.ts"], // @remap-prod-remove-line
+      "@reduxjs/toolkit": [
+        "src/index.ts"
+      ], // @remap-prod-remove-line
+      "@reduxjs/toolkit/react": [
+        "src/react/index.ts"
+      ], // @remap-prod-remove-line
+      "@reduxjs/toolkit/query": [
+        "src/query/index.ts"
+      ], // @remap-prod-remove-line
+      "@reduxjs/toolkit/query/react": [
+        "src/query/react/index.ts"
+      ], // @remap-prod-remove-line
       // for type imports in tests only
-      "@reduxjs/toolkit/dist/*": ["src/*"], // @remap-prod-remove-line
+      "@reduxjs/toolkit/dist/*": [
+        "src/*"
+      ], // @remap-prod-remove-line
       // for type imports in tests only
```
These moved comments might prove problematic in some tests

EskiMojo14 replied:
prettier moment - will fix

EskiMojo14 committed:
rename KeysMatching and fix formatting on tsconfig

EskiMojo14 committed:
fix docs

EskiMojo14 committed:
remove unhappy line

EskiMojo14 committed:
Add custom creator test

EskiMojo14 committed:
use matcher to avoid slice hack

EskiMojo14 committed:
use selectSlice because we can

EskiMojo14 committed:
add specific error message when reducer definition type is undefined

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
remove underscore

EskiMojo14 committed:
update errors.json

EskiMojo14 committed:
remove unnecessary rename

EskiMojo14 committed:
add condition creator test

EskiMojo14 committed:
check promise status synchronously

EskiMojo14 committed:
use resolves chain

EskiMojo14 committed:
shuffle creator locations

EskiMojo14 committed:
add multireducer creator example

EskiMojo14 committed:
use symbols for reducer types, for clarity

aryaemami59 committed:
Modify `BuildCreateSliceConfig` to allow passing `asyncThunk` by default and prevent passing of `reducer` and `reducerWithPrepare`

  - Changed `BuildCreateSliceConfig` to allow passing `asyncThunk` without sacrificing autocompletion.
  - Implemented type constraints to disallow usage of `reducer` and `reducerWithPrepare`, removing them from autocomplete suggestions.

aryaemami59 committed:
Fix typo from `reducerWithPrepare` -> `preparedReducer`

EskiMojo14 committed:
expose getInitialState to reducer creators

EskiMojo14 committed:
tweak creators type to prevent autocomplete of banned keys

EskiMojo14 committed:
change default back

EskiMojo14 committed:
pagination and history creators

EskiMojo14 committed:
add PreparedCaseReducerDefinition

EskiMojo14 committed:
separate reducer factory and prepare for more flexibility

EskiMojo14 committed:
rename prepare and add withoutPayload prepare

EskiMojo14 committed:
Add checking/autocomplete for asyncThunk creator key

aryaemami59 committed:
Merge branch 'create-slice-creators' of https://github.com/reduxjs/redux-toolkit into create-slice-creators

EskiMojo14 committed:
export reducer creators and reuse handlers

EskiMojo14 committed:
Properly extract possible reducer definitions for a given handler

EskiMojo14 committed:
prevent overwriting handlers for reducer and reducerWithPrepare

EskiMojo14 committed:
remove unnecessary generic

aryaemami59 committed:
Merge branch 'create-slice-creators' of https://github.com/reduxjs/redux-toolkit into create-slice-creators

aryaemami59 committed:
Fix minor type issues inside `BuildCreateSliceConfig`

aryaemami59 committed:
Add some type tests for `createSlice`

  - Add type tests to make sure the default `createSlice` does not allow `create.asyncThunk()` but should allow `create.reducer()` and `create.preparedReducer()`

EskiMojo14 committed:
Merge pull request #4024 from aryaemami59/create-slice-creators

Modify `BuildCreateSliceConfig` to allow passing `asyncThunk` by default and prevent passing of `reducer` and `reducerWithPrepare`

EskiMojo14 committed:
only assign handler if one exists

EskiMojo14 committed:
Begin writing documentation

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
minor tweaks

EskiMojo14 committed:
Derive RegisteredReducerType from keys of SliceReducerCreators, removing need for ReducerTypes.

EskiMojo14 committed:
rm unused util

EskiMojo14 committed:
find definitions recursively regardless of depth

EskiMojo14 committed:
add default CreatorMap for ReducerCreators type

EskiMojo14 committed:
add ReducerNamesOfType util

EskiMojo14 committed:
add Typescript section

EskiMojo14 committed:
use ReducerNamesOfType more

EskiMojo14 committed:
Batched creator test

EskiMojo14 committed:
define -> create

EskiMojo14 committed:
add "single definition" example

EskiMojo14 committed:
add pagination example

EskiMojo14 committed:
Add "other" example

EskiMojo14 committed:
create must be a function

EskiMojo14 committed:
corrections

EskiMojo14 committed:
Add ReducerCreatorEntry type to streamline creator registry

EskiMojo14 committed:
simplify default

EskiMojo14 committed:
search attached methods for possible definitions too

EskiMojo14 committed:
test attached method returning definition

EskiMojo14 committed:
use ReducerCreatorEntry type in docs

EskiMojo14 committed:
a word

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Use createAppSlice name consistently

EskiMojo14 committed:
Omit creator key entirely if value is never

EskiMojo14 committed:
Add note about compatible state.

EskiMojo14 committed:
Avoid using IfMaybeUndefined

EskiMojo14 committed:
guard against symbol types

EskiMojo14 committed:
guard against more symbols

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
blurb it up

EskiMojo14 committed:
slight tweak

EskiMojo14 committed:
remove extra tests now they're in the docs

EskiMojo14 committed:
provide more context on buildCreateSlice

EskiMojo14 committed:
move buildCreateSlice after examples

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
begin restructuring

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
fix entity slice enhancer test

EskiMojo14 committed:
fix createSlice test

EskiMojo14 committed:
fix 4.7 issue

EskiMojo14 committed:
ensure only asyncThunkCreator is passed to asyncThunk key

EskiMojo14 committed:
rewrite some blurbs

EskiMojo14 committed:
add reserved name warning

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
run prettier

EskiMojo14 committed:
Added entity methods creator

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
test plurals

EskiMojo14 committed:
guard against empty string and "fix" createSlice type

EskiMojo14 committed:
fix id in test

EskiMojo14 committed:
extract config to variable

EskiMojo14 committed:
fix plurals

EskiMojo14 committed:
add entityMethodsCreator to docs

EskiMojo14 committed:
update errors.json

EskiMojo14 committed:
ensure consistency between imports and module augmentation

EskiMojo14 committed:
remove unnecessary annotation

EskiMojo14 committed:
golf

EskiMojo14 committed:
match title

EskiMojo14 committed:
move collision warning

EskiMojo14 committed:
move ternary

EskiMojo14 committed:
add defaultplural type

EskiMojo14 committed:
import reducerCreator directly to avoid reliance on `this`

EskiMojo14 committed:
unused import

EskiMojo14 committed:
differ internal typing for singular and plural to ensure a match

EskiMojo14 committed:
add comment

EskiMojo14 committed:
prevent any when calling entityMethodsCreator.create

EskiMojo14 committed:
improve internal type safety

EskiMojo14 committed:
Fix type checking in reducers field by making callback into generic

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
add write permissions

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
adjust tests

EskiMojo14 committed:
adjust test more

EskiMojo14 committed:
avoid satisfies expression

EskiMojo14 committed:
tweak createSlice logic

EskiMojo14 committed:
try adding a pure

EskiMojo14 committed:
pure more of the things

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
remove unnecessary pures

EskiMojo14 committed:
fix tree shaking by comparing type instead of identity

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
remove todo

EskiMojo14 committed:
Move type test to type-d file

EskiMojo14 committed:
strongly type action types

EskiMojo14 committed:
remove defaults from SliceReducerCreator

EskiMojo14 committed:
pass name and state to SliceReducerCreators when getting final actions/case reducers type

EskiMojo14 committed:
use CreatorCaseReducers type in more places

EskiMojo14 committed:
cast harder

EskiMojo14 committed:
add notes about type parameters

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
remove unused barrel file

EskiMojo14 committed:
fix imports in tests

EskiMojo14 committed:
add named selectors

EskiMojo14 committed:
call with whole action

EskiMojo14 committed:
ensure selectors are computed

EskiMojo14 committed:
remove unused PURE

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

markerikson commented on packages/toolkit/src/createSlice.ts:
```diff
@@ -585,17 +849,167 @@ function getType(slice: string, actionKey: string): string {
+  },
+  handle({ type, reducerName }, reducer, context) {
+    context
+      .addCase(type, reducer as any)
```
:hand: this seems like a common sequence to do. Worth building it in?

EskiMojo14 replied:
hmm, what would it be called?
currently a custom creator can just import reducerCreator and use it directly:

```ts
reducerCreator.handle(
  { type, reducerName },
  reducerCreator.create(reducer),
  context,
);
```

I do this in my history-adapter creator: https://github.com/EskiMojo14/history-adapter/blob/main/src/creator.ts#L122-L126

markerikson commented on packages/toolkit/src/createSlice.ts:
```diff
@@ -673,30 +1089,51 @@ export function buildCreateSlice({ creators }: BuildCreateSliceConfig = {}) {
+          reducerName,
+          type: getType(name, reducerName),
+        }
+        if ('reducer' in reducerDefinition) {
```
:hand: silly byte-shaving - can we do `const handler = 'reducer' in reducerDefinition ? preparedReducerCreator : reducerCreator`, and then unconditionally do `handler(reducerDetails, reducerDefinition, contextMethods)` ?

EskiMojo14 replied:
ya - originally this made typescript happier but since we're `as any`ing anyway, might as well

markerikson commented on packages/toolkit/src/entities/slice_creator.ts:
:hand: For readability purposes, can you break all the entity creator stuff out into a separate PR?  I'd like this PR to focus on the core mechanism itself.

EskiMojo14 replied:
done: https://github.com/reduxjs/redux-toolkit/pull/4223

markerikson commented on packages/toolkit/src/createSlice.ts:
```diff
@@ -25,19 +30,317 @@ import type {
+      }
+    }
+  >
+  [ReducerType.asyncThunk]: ReducerCreatorEntry<
```
:question: does it make any sense to break out the async thunk creator code and definitions into a separate file?  It'd still be built into RTK, but if there's already the ability to define custom creators anyway, it feels like this is both an example of a custom creator and something that could be pulled out from the `createSlice.ts` file

EskiMojo14 replied:
done - SliceReducerCreators part needed to stay in createSlice.ts, but the rest has been moved

EskiMojo14 commented on packages/toolkit/src/createReducer.ts:
```diff
@@ -152,14 +152,7 @@ export function createReducer<S extends NotFunction<any>>(
   let [actionsMap, finalActionMatchers, finalDefaultCaseReducer] =
     executeReducerBuilderCallback(mapOrBuilderCallback)
 
-  // Ensure the initial state gets frozen either way (if draftable)
```
yep - currently we get `getInitialState` by building the reducer, but in order to expose it to the handler context it needs to be moved out so we can have it before the reducer is built

EskiMojo14 commented on packages/toolkit/src/createSlice.ts:
```diff
@@ -585,17 +849,167 @@ function getType(slice: string, actionKey: string): string {
+    return Object.assign(
+      {
+        // hack so the wrapping function has the same name as the original
+        // we need to create a wrapper so the `reducerDefinitionType` is not assigned to the original
```
we need to assign the _reducerDefinitionType brand to the reducer we return, so it can be passed to the right handler. however, we probably don't want to assign it to the function reference passed to the creator. this creates a new function reference with the same name as the one passed in (for trace purposes) and assigns the brand to that instead

EskiMojo14 commented on docs/api/createSlice.mdx:
````diff
@@ -662,3 +506,1305 @@ console.log(counter.actions.decrement.type)
+export const { addTodo, deleteTodo, fetchTodo } = todosSlice.actions
+```
+
+#### RTK Creators
````
for reference, [this discussion](https://github.com/reduxjs/redux-toolkit/discussions/4020#discussioncomment-8057488) was the feedback I was trying to account for - currently we just throw `buildCreateSlice` out there without really explaining what it's doing, whereas ideally we'd explain how `buildCreateSlice` works *before* introducing our custom creators (`asyncThunk`, `entityMethods`) that you need `buildCreateSlice` for

markerikson replied:
My thought was that the API reference page first focuses on introducing the callback syntax with just the built-in capabilities (`create.reducer/preparedReducer`), and briefly mentions "hey look you can customize this too" and points to the additional details.

EskiMojo14 replied:
yeah, my concern is that i don't want to bury create.asyncThunk or create.entityMethods too much, i think they'll be widely useful but only if known about

EskiMojo14 committed:
byte shaving

EskiMojo14 committed:
expose sliceName to reducer handler

EskiMojo14 committed:
split asyncThunkCreator into its own file

EskiMojo14 committed:
fix typo

EskiMojo14 committed:
fix type imports

EskiMojo14 committed:
split type test

EskiMojo14 committed:
move module augmentation back to createSlice

EskiMojo14 committed:
Rework mapped type to keep homogenous mapping

EskiMojo14 commented on packages/toolkit/src/createSlice.ts:
```diff
@@ -460,33 +742,18 @@ type SliceActionType<
+  CaseReducers extends CreatorCaseReducers<State>,
   SliceName extends string,
-> = {
-  [Type in keyof CaseReducers]: CaseReducers[Type] extends infer Definition
```
you [can](https://github.com/reduxjs/redux-toolkit/pull/3894/commits/5f5fe7952630acc7b28de4d0e371522a23922cba), but you lose the ergonomics of `ReducerNamesOfType`. You end up always having to do
```ts
{
  [ReducerName in keyof CaseReducers]: CaseReducers[ReducerName] extends ReducerDefinition<typeof myCustomType> ? MyActualLogicHere : never
}
```

Definitions that aren't recognised by any handlers also get mapped to `unknown` instead of not being exposed at all: 
```ts
const slice = createSlice({
  name,
  initialState,
  reducers: {
    foo: "banana!"
  }
})

slice.actions.foo // previously not exposed as a key, now is `unknown`
```

the types are also having to do more work because there's 1) more intersections and 2) we're converting any `never`s into `unknown` ("foo" & never = never, "foo" & unknown = "foo")

EskiMojo14 replied:
removed `ReducerNamesOfType` and updated docs with conditional needed instead

EskiMojo14 committed:
there are no entity methods creators in ba sing se

EskiMojo14 committed:
reset more entity files

EskiMojo14 committed:
again

EskiMojo14 committed:
fix test

EskiMojo14 committed:
remove ReducerNamesOfType

EskiMojo14 commented on docs/api/createSlice.mdx:
````diff
@@ -134,14 +134,18 @@ const todosSlice = createSlice({
 ```
 
-### The `reducers` "creator callback" notation
+:::tip Slice creators and the `reducers` "creator callback" notation
````
yeah, the idea was to have a short "hey here's a preview of cool stuff, go here for more details" so `create.asyncThunk` wouldn't get buried

EskiMojo14 committed:
bump version reference

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 commented on docs/api/createSlice.mdx:
```diff
@@ -662,3 +506,1305 @@ console.log(counter.actions.decrement.type)
+
+In version v2.2.0 (TODO), we introduced a system for including your own creators.
+
+The below documentation will cover how to write your own creators, and how to use them with `createSlice`.
```
should create.asyncThunk and create.entityMethods be documented here or on the usage guide?

EskiMojo14 committed:
add usage page

EskiMojo14 committed:
move creator callback section back and cut and paste slice creators section to usage guide

EskiMojo14 committed:
take titles down a step

EskiMojo14 committed:
a word

EskiMojo14 committed:
fix links

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
omit any from extracted definitions

EskiMojo14 committed:
cast more anys

EskiMojo14 committed:
rename context types

EskiMojo14 committed:
remove first argument for exposeAction and exposeCaseReducer, and always use reducerName

EskiMojo14 committed:
add tests for context errors

EskiMojo14 committed:
add note about only calling methods once

EskiMojo14 committed:
expose selectSlice, name and reducerPath in context

EskiMojo14 committed:
expose Name and ReducerPath in types, and add test for selectSlice

EskiMojo14 committed:
move reducerPath to reducerDetails

EskiMojo14 committed:
simplify selectSlice types

EskiMojo14 committed:
update docs definitions to match

EskiMojo14 committed:
update errors.json

EskiMojo14 committed:
update docs with new options

EskiMojo14 committed:
avoid satisfies for type test

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
add no-transpile to snippet

EskiMojo14 committed:
test symbol errors

EskiMojo14 committed:
golf

EskiMojo14 committed:
make sure each creator symbol has a description

EskiMojo14 committed:
simplify GetCaseReducers

EskiMojo14 committed:
use utility more

EskiMojo14 committed:
unsimplify GetCaseReducers because it broke everything

EskiMojo14 committed:
golf

EskiMojo14 committed:
remove unnecessary cast

EskiMojo14 committed:
expand upon creator docs further

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
move async thunk creator module augmentation

EskiMojo14 committed:
try something different

EskiMojo14 committed:
tweak example

EskiMojo14 committed:
add note about exposing multiple values per reducer def

EskiMojo14 committed:
add note about addMatcher order

EskiMojo14 committed:
limit recursion depth of definition extraction

EskiMojo14 committed:
rm unused type

EskiMojo14 committed:
guard against strange behaviour of OverloadedReturnType in pre 5.3 versions

EskiMojo14 committed:
condense OverloadedReturnType since first conditional is always hit

EskiMojo14 committed:
remove note re: addMatcher order as it doesn't apply

EskiMojo14 committed:
remove requirement for create to be a function

EskiMojo14 committed:
test non-function creator

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
fix type portability examples

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
switch back to an interface

EskiMojo14 committed:
avoid relative declaration

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
Merge branch 'master' into create-slice-creators

EskiMojo14 committed:
export async thunk slice config type

EskiMojo14 committed:
separate creators and the actions/caseReducers they expose, to avoid footgun of CaseReducers seeming available to creator type

EskiMojo14 committed:
update docs

EskiMojo14 committed:
only pass State to SliceReducerCreators
